### PR TITLE
fix(polling/state): prevent duplicates in repeatable requests list

### LINKED ISSUE #8390: `InFlightLongPollingActivateJobsRequestsState#activeRequestsToBeRepeated` contains duplicates
**Describe the bug**

See description in #8310:

**Scenario 1: Duplicates in `activeRequestsToBeRepeated` when other requests succeeds**

Given three activate requests `x`, `y`, and `z`:

1. send request `z` to broker -> current `activeRequests = [ z ]` and `activeRequestsToBeRepeated = [ ]`
2. send request `y` to broker -> current `activeRequests = [ z, y ]` and `activeRequestsToBeRepeated = [ ]`
3. send request `x` to broker -> current `activeRequests = [ z, y, x ]` and `activeRequestsToBeRepeated = [ ]`
4. request `z` completes with at least one activated job -> reset failed attempts -> current `activeRequests = [ y, x ]` and `activeRequestsToBeRepeated = [ y, x ]`
5. request `y` completes with at least one activated job -> reset failed attempts -> current `activeRequests = [ x ]` and `activeRequestsToBeRepeated = [ x, x ]`

**Scenario 2: Duplicates in `activeRequestsToBeRepeated` on multiple notifications**

Given is a request `x`:

1. send request `x` to broker -> current `activeRequests = [ x ]` and `activeRequestsToBeRepeated = [ ]`
2. Broker notifies the gateway about new available jobs -> reset failed attempts -> current `activeRequests = [ x ]` and `activeRequestsToBeRepeated = [ x ]`
3. Again, Broker notifies the gateway about new available jobs -> reset failed attempts -> current `activeRequests = [ x ]` and `activeRequestsToBeRepeated = [ x, x ]`

**Scenario 3: Another request wins**

Given are requests `[z1, z2, z3, ...., zn]` and `x`:

1. send requests `[z1, z2, z3, ...., zn]` and `x` -> current `activeRequests = [ z1, z2, z3, ...., zn, x ] and `activeRequestsToBeRepeated = [ ]`
2. request `z1` completes with at least one activated job -> reset failed attempts -> current `activeRequests = [  z2, z3, ...., zn, x ]` and `activeRequestsToBeRepeated = [  z2, z3, ...., zn, x ]`
3. request `x` completes without any jobs -> retry request -> current `activeRequests = [  z2, z3, ...., zn, x ]` and `activeRequestsToBeRepeated = [  z2, z3, ...., zn ]`
4. request `z2` completes with at least one activated job -> reset failed attempts -> current `activeRequests = [  z3, ...., zn, x ]` and `activeRequestsToBeRepeated = [  z3, ...., zn, x ]`
3. request `x` completes without any jobs -> retry request -> current `activeRequests = [  z3, ...., zn, x ]` and `activeRequestsToBeRepeated = [  z2, z3, ...., zn ]`
4. ...

In all scenarios, the queue `activeRequestsToBeRepeated` contains the request `x` at least once before the response arrives. In a nutshell, as long as the failed attempts are lower than the configured threshold and the request `x` is present in the queue of repeatable requests, the request `x` is executed in a loop when it responds with any activated job.

**Expected behavior**
* `InFlightLongPollingActivateJobsRequestsState#activeRequestsToBeRepeated` does not contain duplicates

**Environment:**
 - Zeebe Version: 1.2.5
 - Configuration: Enabled Long Polling

---

is depended on by #8310

### PR DESCRIPTION
**Pull request by romansmirnov**
## Related issues

relates #8310
closes #8390

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by romansmirnov** [comment]
I know, I know... using `HashSet` without implementing `hashCode()` and `equals()` in `LongPollingActivateJobsRequest` 😄 But none of the existing given properties (each on its own and all together) helps to identify a request clearly. For example, a replicated worker shares the same configuration which would result in multiple separate requests with the same request property values. So that, the requests could not be distinguished anymore as they would be considered as equal.

Possible Solutions:
1. Continue using `HashSet` but introduce a unique identifier for each incoming request per job type. The ID could be used to implement `equals()` and `hashCode()`.
2. Keep the type `LinkedList` but check if the request is already contained by the list
```
activeRequests.forEach((r) -> {
  if (!activeRequestsToBeRepeated.contains(r)) {
    activeRequestsToBeRepeated.add(r);
  }
});
```
3. Accept the duplicates as they all will be removed when the request times out (but the request might be executed multiple times unnecessarily)
4. ...

**Comment by romansmirnov** [comment]
Dear Reviewer,

Please consider the following:

* Introduce a `requestId` property in `LongPollingActivateJobsRequest` which allows to clearly identify a request. Additionally, that allows to implement `#hashCode()` and `#equals()` accordingly. Otherwise, the existing properties do not allow to differentiate them properly.
* The `requestId` is not persisted, meaning, after a Gateway restart the `requestId` starts again with `1`.

Also, the PR might be in conflict with the PR #8389 when merging it.

File: gateway/src/test/java/io/camunda/zeebe/gateway/api/job/LongPollingActivateJobsTest.java [hunk 1]
```diff
@@ -510,6 +521,52 @@ public void shouldReturnJobsIfSomeBrokersHaveJobsWhileOthersReturnResourceExhaus
...
+  public void shouldRepeatRequestOnlyOnce() throws Exception {
+    // given
+    // the first three requests activates jobs
+    final var firstRequest = getLongPollingActivateJobsRequest();
+    final var secondRequest = getLongPollingActivateJobsRequest();
+    final var thirdRequest = getLongPollingActivateJobsRequest();
+    // the last request does not activate any jobs
+    final var fourthRequest = getLongPollingActivateJobsRequest();
+
+    final var allRequestsSubmittedLatch = new CountDownLatch(1);
```
Comment (romansmirnov) [inline gateway/src/test/java/io/camunda/zeebe/gateway/api/job/LongPollingActivateJobsTest.java:534]: Dear Reviewer, just FYI: I consciously decided to use the `CountDownLatch` to ensure that all requests are submitted to the actor initially. As an alternative, I considered using the `ControlledActorScheduler` but the test was too flaky with the controlled scheduler. With the latch, the test is reliable.

**Review by lenaschoenburg — CHANGES_REQUESTED** [review]
:x:  I think introducing a request id is not necessary because we can just rely on identity to check for duplicates in `activeRequestsToBeRepeated`. We create new instances of `LongPollingActivateJobsRequest` only once. 

That means `activeRequestsToBeRepeated` could be: 
```
private final Set<LongPollingActivateJobsRequest> activeRequestsToBeRepeated =
      Collections.newSetFromMap(new IdentityHashMap<>());
```
Unless we want to use the request id for more than just checking duplicates I think we should not introduce it here and instead rely on object identity. 

WDYT?

**Comment by npepinpe** [comment]
Regarding identity, what you say sounds reasonable - as long as we only create requests once and each instance is meant to be unique, it should be OK to simply use the default `equals`. I can't really speak on his intentions and reasoning, but I can imagine two reasons why Roman would've implemented a custom one however (which would require an ID as otherwise there's no way to guarantee we can distinguish requests):

1. What if that constraint changes? It's not a specific constraint of the class that it must be instantiated only once per request, so it's not too crazy to think someone might not respect it.
1. What if later on, someone thinks: "hey, what if we aggregated similar requests that share the same properties/subset to reduce load on the broker? we could implement a custom equals/hashCode and put that in a set so they're automatically deduplicated per job type". This may break the current expectations of uniqueness, and maybe it wouldn't be immediately obvious what happened, or easy to diagnose.

That's just OTOH though - my proposal would be to accept with the ID, and open a follow up to discuss it with Roman when he's back - there's no rush here, and removing the ID or taking a different approach in the long run is easy to do. Let me know what you think :+1:

**Review by lenaschoenburg — APPROVED** [review]
I rebased to resolve conflicts with the other gateway/long-polling related PRs.

bors r+

**Pull request merged**